src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -554,32 +565,74 @@ private Task ProcessHeadersFrameAsync<TContext>(IHttpApplication<TContext> appli
-                    ServiceContext = _context.ServiceContext,
-                    ConnectionFeatures = _context.ConnectionFeatures,
-                    MemoryPool = _context.MemoryPool,
-                    LocalEndPoint = _context.LocalEndPoint,
-                    RemoteEndPoint = _context.RemoteEndPoint,
-                    StreamLifetimeHandler = this,
-                    ClientPeerSettings = _clientSettings,
-                    ServerPeerSettings = _serverSettings,
-                    FrameWriter = _frameWriter,
-                    ConnectionInputFlowControl = _inputFlowControl,
-                    ConnectionOutputFlowControl = _outputFlowControl,
-                    TimeoutControl = TimeoutControl,
-                });
-
-                _currentHeadersStream.Reset();
+                _currentHeadersStream = CreateStream(application);
+
                 _headerFlags = _incomingFrame.HeadersFlags;
 
                 var headersPayload = payload.Slice(0, _incomingFrame.HeadersPayloadLength); // Minus padding
                 return DecodeHeadersAsync(_incomingFrame.HeadersEndHeaders, headersPayload);
             }
         }
 
+        private Http2Stream CreateStream<TContext>(IHttpApplication<TContext> application)
+        {
+            Http2Stream<TContext> stream = null;
+
+            lock (_streamPool)
+            {
+                if (_pooledStreamCount > 0)
+                {
+                    _pooledStreamCount--;
+                    stream = (Http2Stream<TContext>)_streamPool[_pooledStreamCount];
+                    stream.Reset();
+                    stream.Initialize(GetHttp2StreamContext());
+                }
+            }
+
+            if (stream == null)
+            {
+                stream = new Http2Stream<TContext>(
+                    application,
+                    GetHttp2StreamContext());
+                stream.Reset();
+            }
+
+            return stream;
+        }
+
+        private Http2StreamContext GetHttp2StreamContext()
+        {
+            return new Http2StreamContext
+            {
+                ConnectionId = ConnectionId,
+                StreamId = _incomingFrame.StreamId,
+                ServiceContext = _context.ServiceContext,
+                ConnectionFeatures = _context.ConnectionFeatures,
+                MemoryPool = _context.MemoryPool,
+                LocalEndPoint = _context.LocalEndPoint,
+                RemoteEndPoint = _context.RemoteEndPoint,
+                StreamLifetimeHandler = this,
+                ClientPeerSettings = _clientSettings,
+                ServerPeerSettings = _serverSettings,
+                FrameWriter = _frameWriter,
+                ConnectionInputFlowControl = _inputFlowControl,
+                ConnectionOutputFlowControl = _outputFlowControl,
+                TimeoutControl = TimeoutControl,
+            };
+        }
+
+        private void ReturnStream(Http2Stream stream)
```

Comment:
If you called Reset on return then you wouldn't root the non-pooled content as long.